crates/subspace-service/src/dsn/import_blocks/root_blocks.rs

```diff
@@ -87,36 +98,68 @@ impl RootBlockHandler {
-                                trace!(
-                                    %peer_id,
-                                    segment_index=root_block.segment_index(),
-                                    "Last root block was obtained."
-                                );
-
-                                return Ok(*root_block);
-                            } else {
-                                debug!(%peer_id, "Last root block was not received.");
-                            }
+                            trace!(
+                                %peer_id,
+                                root_blocks_number=%root_blocks.len(),
+                                "Last root block request succeeded."
+                            );
+
+                            peer_blocks.insert(peer_id, root_blocks);
                         }
                         Err(error) => {
                             debug!(%peer_id, ?error, "Last root block request failed.");
                         }
                     };
                 }
-                Err("No more peers for root blocks.".into())
             }
             Err(err) => {
                 warn!(?err, "get_closest_peers returned an error");
 
-                Err(err.into())
+                return Err(err.into());
+            }
+        }
+
+        // TODO: Consider adding attempts to increase the initial peer set.
+        if peer_blocks.len() < ROOT_BLOCK_CONSENSUS_MIN_SET {
+            return Err(format!(
+                "Root block consensus failed: not enough peers ({}).",
+                peer_blocks.len()
+            )
+            .into());
+        }
+
+        // Calculate votes
+        let mut root_block_score: BTreeMap<Blake2b256Hash, u64> = BTreeMap::new();
+        let mut root_block_dict: BTreeMap<Blake2b256Hash, RootBlock> = BTreeMap::new();
+
+        for (_peer_id, root_blocks) in peer_blocks {
+            // TODO: check root_blocks variable for duplicates and add peer ban
+            for root_block in root_blocks {
+                root_block_score
+                    .entry(root_block.hash())
+                    .and_modify(|val| *val += 1)
+                    .or_insert(1);
+                root_block_dict
+                    .entry(root_block.hash())
+                    .or_insert(root_block);
+            }
+        }
+
+        // Sort the collection to get highest blocks first.
+        let mut root_blocks = root_block_dict.values().collect::<Vec<_>>();
+        root_blocks.sort_by_key(|rb| Reverse(rb.segment_index()));
+
+        for root_block in root_blocks {
+            let score = root_block_score
+                .get(&root_block.hash())
+                .expect("Must be present because of the manual adding.");
+
+            if *score >= ROOT_BLOCK_CONSENSUS_THRESHOLD {
+                return Ok(*root_block);
             }
         }
+
+        Err("Root block consensus failed: can't pass the threshold.".into())
```

Comment:
This doesn't seem like a fatal error, we should be able to try again a few times and maybe succeed eventually.

Reply:
But why not right away? IT doesn't seem like a difficult change worth shipping clearly inferior version just to see that we do indeed to implement retries.